### PR TITLE
f64 kernels

### LINKED ISSUE #224: Double precision tensors
Hey, I think this crate could be very useful outside of basic machine learning applications. However, in scientific calculations double precision floats are the norm. In a previous Issue #12 you mentioned that you stopped the work on making `Tensor`s generic over their `DType`. 

Is that still the intended way forward, or are you interested to have tensor operations either generic or specifically for `f64`?

### PR DESCRIPTION
**Pull request by coreylowman**
- Adds `num-trait` dependency
- Uses `num_traits::Float` in all Cpu kernels
- impls Cuda kernels for both f32 & f64 for now.
- Adds `test-f64` feature flag for dev
- Adds `dev.build_module` back to make specifying dtype easier

Resolves #224

### CODE REVIEW

File: src/tensor/cpu/allocate.rs [hunk 1]
```diff
@@ -64,22 +66,19 @@ impl<E: Unit> ZeroFillStorage<E> for Cpu {
     }
 }
 
-impl OnesTensor<f32> for Cpu {
-    fn try_ones_like<S: HasShape>(
-        &self,
-        src: &S,
-    ) -> Result<Tensor<S::Shape, f32, Self>, Self::Err> {
-        let storage = StridedArray::try_new_with(*src.shape(), 1.0)?;
+impl<E: Unit + One> OnesTensor<E> for Cpu {
```
Comment (coreylowman) [inline src/tensor/cpu/allocate.rs:69]: @nkoppel FYI related to your One change - we could pull out the num_traits dependency into a smaller PR that just adds it for this for now (probably best to stage this giant PR anyway).

File: src/tensor_ops/min_to/min_to.cu [hunk 2]
```diff
@@ -10,15 +10,38 @@ __device__ __forceinline__ float atomicMinf(float * addr, float value) {
...
 
+// atomicMax is not implemented for double,
+// solution copied https://stackoverflow.com/questions/17399119/how-do-i-use-atomicmax-on-floating-point-values-in-cuda
+__device__ __forceinline__ double atomicMinf(double * addr, double value) {
+    if (signbit(value)) {
+        return __uint_as_double(atomicMax((unsigned int *)addr, __double_as_uint(value)));
+    } else {
+        return __int_as_double(atomicMin((int *)addr, __double_as_int(value)));
+    }
+}
```
Comment (coreylowman) [inline src/tensor_ops/min_to/min_to.cu:21]: This & the max version currently don't work because these intrinsics don't work.

There are the __double2hiint and __double2loint, but then the comparison operations wouldn't work

Comment (ViliamVadocz) [reply src/tensor_ops/min_to/min_to.cu:21]: Why not convert to `long long int` instead using [__double_as_longlong](https://docs.nvidia.com/cuda/cuda-math-api/group__CUDA__MATH__INTRINSIC__CAST.html#group__CUDA__MATH__INTRINSIC__CAST_1geb72123ab9c35d88842baf34140a8ca6) and [__longlong_as_double](https://docs.nvidia.com/cuda/cuda-math-api/group__CUDA__MATH__INTRINSIC__CAST.html#group__CUDA__MATH__INTRINSIC__CAST_1g354a51915a0c6465ac5e5c844724d987)?

Comment (coreylowman) [reply src/tensor_ops/min_to/min_to.cu:21]: Oh! I totally missed those somehow lmao. Great find!!

File: src/tensor/cpu/allocate.rs (hunk 1 above)
Comment (nkoppel) [reply src/tensor/cpu/allocate.rs:69]: Unfortunately, One is not implemented for bool, so my current implementation is probably the best we can do.

File: src/nn/dropout.rs [hunk 3]
```diff
@@ -60,12 +60,12 @@ impl<const N: usize, S: Shape, E: Dtype, D: Device<E>> Module<Tensor<S, E, D, No
     }
 }
 
-impl<const N: usize, S: Shape, E: Dtype, D: Device<E>> ModuleMut<Tensor<S, E, D, OwnedTape<D>>>
+impl<const N: usize, S: Shape, D: Device<f32>> ModuleMut<Tensor<S, f32, D, OwnedTape<D>>>
```
Comment (coreylowman) [inline src/nn/dropout.rs:63]: Revert this change? should be generic over dtype

File: src/nn/embedding.rs [hunk 4]
```diff
@@ -47,10 +52,16 @@ pub struct Embedding<const VOCAB: usize, const DIM: usize, E: Dtype, D: DeviceSt
     pub weight: Tensor<Rank2<VOCAB, DIM>, E, D>,
 }
 
-impl<const VOCAB: usize, const DIM: usize, const SEQ: usize, D: Device<f32>, T: Tape<D>>
-    Module<Tensor<Rank1<SEQ>, usize, D, T>> for Embedding<VOCAB, DIM, f32, D>
+impl<
+        const VOCAB: usize,
+        const DIM: usize,
+        const SEQ: usize,
+        E: Dtype,
```
Comment (coreylowman) [inline src/nn/embedding.rs:59]: style

File: src/nn/embedding.rs [hunk 5]
```diff
@@ -158,7 +170,7 @@ mod tests {
         let dev: TestDevice = Default::default();
 
         let model = Embedding {
-            weight: dev.tensor(W),
+            weight: dev.tensor(W.map(|row| row.map(TestDtype::from))),
```
Comment (coreylowman) [inline src/nn/embedding.rs:173]: change definition of W to use TestDtype instead of f32

File: src/nn/linear.rs [hunk 6]
```diff
@@ -47,45 +52,47 @@ pub struct Linear<const I: usize, const O: usize, E: Dtype, D: DeviceStorage> {
...
+    ResetParams<D, E> for Linear<I, O, E, D>
 {
     fn try_reset_params(&mut self) -> Result<(), D::Err> {
-        let bound: f32 = 1.0 / (I as f32).sqrt();
-        let distr = rand_distr::Uniform::new(-bound, bound);
-        self.weight.try_fill_with_distr(distr)?;
-        self.bias.try_fill_with_distr(distr)?;
+        let bound: E = E::ONE / E::from_usize(I).unwrap().sqrt();
+        self.weight
+            .try_fill_with_distr(rand_distr::Uniform::new(-bound, bound))?;
```
Comment (coreylowman) [inline src/nn/linear.rs:85]: style

File: src/nn/linear.rs [hunk 7]
```diff
@@ -189,11 +201,11 @@ mod tests {
         let dev: TestDevice = Default::default();
 
         let model = Linear {
-            weight: dev.tensor(W),
-            bias: dev.tensor(B),
+            weight: dev.tensor(W.map(|row| row.map(TestDtype::from))),
```
Comment (coreylowman) [inline src/nn/linear.rs:204]: change type of W

File: src/nn/module.rs [hunk 8]
```diff
@@ -52,8 +52,18 @@ pub trait BuildOnDevice<D: DeviceStorage, E: Dtype> {
     }
 }
 
+pub trait DeviceBuildExt: DeviceStorage {
```
Comment (coreylowman) [inline src/nn/module.rs:55]: Add docstring & update readme/examples to use this

File: src/nn/npz_impls.rs [hunk 9]
```diff
@@ -206,17 +206,23 @@ impl<T: LoadFromNpz> LoadFromNpz for AddInto<T> {
...
 
 #[cfg(feature = "nightly")]
-impl<const M: usize, const H: usize, const F: usize, const L: usize, D: CopySlice<f32>> SaveToNpz
-    for TransformerDecoder<M, H, F, L, f32, D>
+impl<
+        const M: usize,
+        const H: usize,
+        const F: usize,
+        const L: usize,
+        E: Dtype + NumpyDtype,
```
Comment (coreylowman) [inline src/nn/npz_impls.rs:214]: style

File: src/nn/npz_impls.rs [hunk 10]
```diff
@@ -247,17 +253,23 @@ impl<const M: usize, const H: usize, const F: usize, D: CopySlice<f32>> LoadFrom
...
 
 #[cfg(feature = "nightly")]
-impl<const M: usize, const H: usize, const F: usize, const L: usize, D: CopySlice<f32>> LoadFromNpz
-    for TransformerDecoder<M, H, F, L, f32, D>
+impl<
+        const M: usize,
+        const H: usize,
+        const F: usize,
+        const L: usize,
+        E: Dtype + NumpyDtype,
```
Comment (coreylowman) [inline src/nn/npz_impls.rs:261]: style

File: src/nn/npz_impls.rs [hunk 11]
```diff
@@ -284,8 +296,14 @@ impl<const M: usize, const H: usize, const F: usize, D: CopySlice<f32>> LoadFrom
...
 
 #[cfg(feature = "nightly")]
-impl<const M: usize, const H: usize, const K: usize, const V: usize, D: CopySlice<f32>> SaveToNpz
-    for MultiHeadAttention<M, H, K, V, f32, D>
+impl<
+        const M: usize,
+        const H: usize,
+        const K: usize,
+        const V: usize,
+        E: Dtype + NumpyDtype,
```
Comment (coreylowman) [inline src/nn/npz_impls.rs:304]: style

File: src/nn/transformer/decoder.rs [hunk 12]
```diff
@@ -278,11 +291,14 @@ mod tests {
         const NUM_HEADS: usize = 6;
         const FF_DIM: usize = 2;
 
-        let decoder =
-            builder::TransformerDecoderBlock::<EMBED_DIM, NUM_HEADS, FF_DIM>::build_on_device(&dev);
+        type Dtype = f32;
```
Comment (coreylowman) [inline src/nn/transformer/decoder.rs:294]: Test out TestDtype since apparently sample_normal is stable over dtype?

File: src/nn/transformer/encoder.rs [hunk 13]
```diff
@@ -198,10 +205,13 @@ mod tests {
         const NUM_HEADS: usize = 3;
         const FF_DIM: usize = 16;
 
-        let encoder =
-            builder::TransformerEncoderBlock::<EMBED_DIM, NUM_HEADS, FF_DIM>::build_on_device(&dev);
+        type Dtype = f32;
```
Comment (coreylowman) [inline src/nn/transformer/encoder.rs:208]: TestDtype

File: src/nn/transformer/mha.rs [hunk 14]
```diff
@@ -83,12 +92,18 @@ impl<const M: usize, const H: usize, const K: usize, const V: usize, D: Device<f
...
 }
 
-impl<const M: usize, const H: usize, const K: usize, const V: usize, D: DeviceStorage>
-    GradientUpdate<D, f32> for MultiHeadAttention<M, H, K, V, f32, D>
+impl<
+        const M: usize,
+        const H: usize,
+        const K: usize,
+        const V: usize,
+        E: Dtype,
```
Comment (coreylowman) [inline src/nn/transformer/mha.rs:100]: style

File: src/nn/transformer/mha.rs [hunk 15]
```diff
@@ -181,32 +198,33 @@ impl<
         const H: usize,
         const K: usize,
         const V: usize,
-        D: Device<f32>,
+        E: Dtype + Float,
```
Comment (coreylowman) [inline src/nn/transformer/mha.rs:201]: Is float necessary here?

File: src/nn/transformer/mha.rs [hunk 16]
```diff
@@ -280,11 +295,13 @@ mod tests {
         const S1: usize = 3;
         const S2: usize = 4;
 
-        let mha = builder::MultiHeadAttention::<M, NUM_HEADS>::build_on_device(&dev);
+        type Dtype = f32;
```
Comment (coreylowman) [inline src/nn/transformer/mha.rs:298]: TestDtype

File: src/optim/adam/adam.cu [hunk 17]
```diff
@@ -51,3 +53,27 @@ extern "C" __global__ void adam_update(
     moment2[i] = v;
     param[i] -= g;
 }
+
+extern "C" __global__ void adam_update_f32(
```
Comment (coreylowman) [inline src/optim/adam/adam.cu:57]: use macros here & with other optimizers

File: src/optim/rmsprop/rmsprop.cu [hunk 18]
```diff
@@ -71,3 +72,27 @@ extern "C" __global__ void rmsprop_update(
     momentum[i] = m;
     param[i] -= g;
 }
+
+extern "C" __global__ void rmsprop_update_f32(
```
Comment (coreylowman) [inline src/optim/rmsprop/rmsprop.cu:76]: macros

File: src/optim/sgd/sgd.cu [hunk 19]
```diff
@@ -56,3 +58,23 @@ extern "C" __global__ void sgd_update(
     velocity[i] = v;
     param[i] -= g;
 }
+
+extern "C" __global__ void sgd_update_f32(
```
Comment (coreylowman) [inline src/optim/sgd/sgd.cu:62]: macros

File: src/tensor/cuda/allocate.rs [hunk 20]
```diff
@@ -68,13 +70,13 @@ where
     }
 }
 
-impl OneFillStorage<f32> for Cuda {
+impl<E: Unit + One> OneFillStorage<E> for Cuda {
```
Comment (coreylowman) [inline src/tensor/cuda/allocate.rs:73]: Use Unit::ONE

File: src/tensor_ops/add/mod.rs [hunk 21]
```diff
@@ -51,8 +51,13 @@ pub trait TryAdd<Rhs = Self>: HasErr {
     fn try_add(self, rhs: Rhs) -> Result<Self, Self::Err>;
 }
 
-impl<S: Shape, E: Dtype, D: Device<E>, LhsTape: Tape<D>, RhsTape: Tape<D>>
-    TryAdd<Tensor<S, E, D, RhsTape>> for Tensor<S, E, D, LhsTape>
+impl<
+        S: Shape,
+        E: Dtype,
```
Comment (coreylowman) [inline src/tensor_ops/add/mod.rs:56]: style

File: src/tensor_ops/choose/choose.cu [hunk 22]
```diff
@@ -46,7 +48,67 @@ extern "C" __global__ void choose_backward(
     unsigned int cond_i = get_strided_index(out_i, num_dims, dims, cond_strides);
 
     auto go = grad_out[out_i];
-    float* out_loc = cond[cond_i] ? grad_lhs + lhs_i : grad_rhs + rhs_i;
+    T* out_loc = cond[cond_i] ? grad_lhs + lhs_i : grad_rhs + rhs_i;
 
     atomicAdd(out_loc, go);
 }
+
+extern "C" __global__ void choose_forward_f32(
```
Comment (coreylowman) [inline src/tensor_ops/choose/choose.cu:56]: macros
